src/lib/Hydra/Controller/User.pm

```diff
@@ -143,6 +157,38 @@ sub doEmailLogin {
 }
 
 
+sub oidc_login :Path('/oidc-login') Args(0) {
+    my ($self, $c) = @_;
+
+    error($c, "Logging in via OIDC is not enabled.", 404) unless $c->config->{enable_oidc_login};
+    my $oidc_client = get_oidc_client($c);
+
+    error($c, q{OIDC state mismatch. Is this a CSRF attack?}) unless $c->session->{oidc_state} && $c->session->{oidc_state} eq $c->request->param("state");
+
+    my $code = $c->request->param("code");
+    my $token_response = $oidc_client->get_access_token(
+        code => $code,
+        redirect_uri => $c->uri_for('/oidc-login'),
+    ) or error($c, $oidc_client->errstr);
+
+    $c->session->{access_token} = $token_response->access_token;
+    $c->session->{expires_at} = time() + $token_response->expires_in;
+    $c->session->{refresh_token} = $token_response->refresh_token;
+
+    my $res = LWP::UserAgent->new->get(
+        $c->config->{oidc_userinfo_uri},
+        Authorization => sprintf(q{Bearer %s}, $token_response->access_token)
+    );
+
+    my $claims = decode_json($res->decoded_content) or error($c, q{Could not decode claims.}, 401);
+
+    error($c, "Email address must be verified.", 401) unless $claims->{email_verified};
+
+    doEmailLogin($self, $c, "oidc", $claims->{email}, $claims->{name} // undef, sprintf("oidc:$claims->{sub}"));
+
```

Comment:
I think it'd be nice to have some role mapping as well because otherwise everyone has no permissions. We use:
```patch
+    my $role_mapping = $c->config->{oidc}->{role_mapping} or ();
+    # Give out roles
+    $c->user->userroles->delete;
+    push(@{$user_data->{groups}}, ("default"));
+    foreach my $group (@{$user_data->{groups}}) {
+        if (defined($role_mapping->{$group})) {
+            my $roles = $role_mapping->{$group};
+            if (ref($roles) eq 'ARRAY') {
+                for my $mapped_role (@{$roles}) {
+                    $c->user->userroles->create({ role => $mapped_role });
+                }
+            } else {
+                $c->user->userroles->create({ role => $roles });
+            }
+        }
+    }
```

Reply:
We discussed this at length (this was what took longest by far), and came to the conclusion that this is a real can of worms, because OIDC is about _authentication_ and _identity_ but not _authorisation_. For now, our recommended approach is to manage permissions manually.